File: app/src/main/java/io/github/sithengineer/motoqueiro/ui/cruising/CruisingModule.java
package io.github.sithengineer.motoqueiro.ui.cruising;

import dagger.Module;
import dagger.Provides;
import io.github.sithengineer.motoqueiro.scope.ActivityScope;
import io.github.sithengineer.motoqueiro.util.CompositeSubscriptionManager;

@Module public class CruisingModule {

  private final CruisingContract.View view;
  private final CruisingNavigator cruisingNavigator;

  public CruisingModule(CruisingContract.View view, CruisingNavigator cruisingNavigator) {
    this.view = view;
    this.cruisingNavigator = cruisingNavigator;
  }

  @Provides @ActivityScope CruisingContract.View provideView() {
    return view;
  }

  @Provides @ActivityScope CruisingNavigator provideNavigator() {
    return cruisingNavigator;
  }

  @Provides @ActivityScope CompositeSubscriptionManager provideCompositeSubscriptionManager() {
    return new CompositeSubscriptionManager();
  }

  @Provides @ActivityScope CruisingContract.Presenter provideCruisingPresenter(
      CruisingContract.View view, CompositeSubscriptionManager subscriptionManager,
      CruisingNavigator navigator) {
    return new CruisingPresenter(view, subscriptionManager, navigator);
  }
}

File: app/src/main/java/io/github/sithengineer/motoqueiro/ui/cruising/CruisingModule.kt
package io.github.sithengineer.motoqueiro.ui.cruising

import dagger.Module
import dagger.Provides
import io.github.sithengineer.motoqueiro.scope.ActivityScope
import io.github.sithengineer.motoqueiro.util.CompositeSubscriptionManager

@Module
class CruisingModule constructor(
    private val view: CruisingContract.View,
    private val cruisingNavigator: CruisingNavigator
) {

  @Provides
  @ActivityScope
  internal fun provideView(): CruisingContract.View {
    return view
  }

  @Provides
  @ActivityScope
  internal fun provideNavigator(): CruisingNavigator {
    return cruisingNavigator
  }

  @Provides
  @ActivityScope
  internal fun provideCruisingPresenter(
      view: CruisingContract.View,
      navigator: CruisingNavigator): CruisingContract.Presenter {
    return CruisingPresenter(view, CompositeSubscriptionManager(), navigator)
  }
}
